Scope renderer to instance

rschamp opened the pull request:
This allows usage without global scope to attach a renderer to the VM. It also provides the ability to have multiple VMs/renderers to be used at once.

Commit:
Scope renderer to instance

This allows usage without global scope to attach a renderer to the VM. It also provides the ability to have multiple VMs/renderers to be used at once.

rschamp commented on src/engine/runtime.js:
```diff
@@ -19,14 +19,21 @@ var defaultBlockPackages = {
+     * Renderer
+     * @type {!RenderWebGL}
+     */
+    this.renderer = renderer;
```
All the things that needed access to the renderer had access to the runtime, so I decided to make it a property of the runtime.  It seems like an ok thing for the runtime to have since it manages the "stuff", but I am open to suggestions about a better place to put this.

tmickel replied:
Having it a property of runtime seems fine to me. But I also think it should be optional/possible to be missing.

tmickel commented on src/index.js:
```diff
@@ -10,16 +10,17 @@ var Blocks = require('./engine/blocks');
+ * @param {!RenderWebGL} renderer Renderer for the VM
  */
-function VirtualMachine () {
+function VirtualMachine (renderer) {
```
Would maybe prefer it I think if this were an optional parameter (`opt_renderer`, `@param {RenderWebGL=}`) or if it was set after instantiation (`vm.attachRenderer`)? I'm imagining we might sometimes want a VM with no renderer.

Commit:
Only expose VirtualMachine to window for bundles

Commit:
Use attachRenderer rather than the constructor

This makes the renderer optional as well as making the order of instantiation flexible.

rschamp commented:
@tmickel I went the `attachRenderer` route which resolved both issues.

I also took out the global scope for the VM to get rid of the remaining globals.  The package that webpack builds still exposes `VirtualMachine` to `window`, but if you import `scratch-vm` nothing gets attached to `window`.
